File: sources/scala/tools/nsc/symtab/classfile/UnPickle.scala
/* NSC -- new scala compiler
 * Copyright 2005 LAMP/EPFL
 * @author  Martin Odersky
 */
// $Id$
package scala.tools.nsc.symtab.classfile {

//import scalac_symtab.Modifiers;
//import scalac_symtab.EntryTags._;
import java.io.PrintStream;
import scala.tools.util.{Position, UTF8Codec};
import java.lang.{Float, Double};

/***************************************************
 * Symbol table attribute format: see EntryTags.java
 */
abstract class UnPickle {

  val global: Global;
  import global._;

  private var classRoot: Symbol = _;
  private var moduleRoot: Symbol = _;
  private var bytes: Array[byte] = _;
  private var bp: int = _;
  private var index: Array[int] = _;
  private var entries: Array[AnyRef] = _;

  def parse(bytes: Array[byte], clazz: Symbol, module: Symbol): unit = {
    warning("cannot yet unpickle: " + clazz.fullNameString('.'));
  }
}}
/*
    this.classRoot = clazz;
    this.moduleRoot = module;
    this.bytes = bytes;
    this.bp = 0;
    if (settings.debug.value) global.log("unpickle " + classRoot + " and " + moduleRoot);
    createIndex();
    if (settings.debug.value) printAll(System.out);
    entries = new Array[Any](index.length);
    for (val i <- Iterator.range(0, index.length)) {
      if (isSymbolEntry(i)) getSymbol(i)
    }
    if (settings.debug.value) global.log("unpickled " + classRoot + ":" + classRoot.rawInfo + ", " + moduleRoot + ":" + moduleRoot.rawInfo);//debug
  }

  def readByte(): int = {
    val x = bytes(bp); bp = bp + 1; x
  }

  def readNat(): int = {
    var b = 0;
    var x = 0;
    do {
      b = readByte();
      x = (x << 7) + (b & 0x7f);
    } while ((b & 0x80) != 0);
    x
  }

  def readLong(len: int): long = {
    var x = 0L;
    var i = 0;
    while (i < len) {
      x = (x << 8) + (readByte() & 0xff);
      i = i + 1
    }
    val leading = 64 - (len << 3);
    x << leading >> leading
  }

  private def at[T](start: int)(op: => T): T = {
    val savedBp = bp;
    bp = start;
    val result = op;
    bp = savedBp;
    result
  }

  private def createIndex(): unit = {
    index = new Array[int](readNat());
    for (val i <- Iterator.range(0, index.length)) {
      index(i) = readByte();
      bp = readNat() + bp
    }
  }

  def isTypeEntry(i: int): boolean = {
    val tag = bytes(index(i));
    (firstTypeTag <= tag && tag <= lastTypeTag) || tag == NOpre;
  }

  def isSymbolEntry(i: int): boolean = {
    val tag = bytes(index(i));
    firstSymTag <= tag && tag <= lastSymTag
  }

  def getName(i: int): Name = {
    var name: Name = entries(i).asInstanceOf[Name];
    if (name == null) {
      at(index(i)) {
        val tag = readByte();
        val len = readNat();
        val cs = new Array[char](len);
        val defined = UTF8Codec.decode(bytes, bp, cs, 0, len);
        name = tag match {
	  case TERMname => newTermName(cs, 0, defined)
	  case TYPEname => newTypeName(cs, 0, defined)
	  case _        => errorBadSignature("");
        }
        entries(i) = name
      }
    }
    name
  }

  def readNameRef(): Name = getName(readNat());

  def getSymbol(i: int): Symbol = {
    var sym: Symbol = entries(i).asInstanceOf[Symbol];
    if (sym == null) {
      at(index(i)) {
        val tag = readByte();
        val end = readNat() + bp;
        tag match {
          case NONEsym =>
            sym = NoSymbol;
            entries(i) = sym
            case EXTref | EXTMODCLASSref =>
              val name = readNameRef();
              val owner =
                if (bp == end) definitions.RootClass
                else { assert(bp < end); readSymbolRef() }
              sym =
                if (name.toTermName == nme.ROOT && owner == NoSymbol) {
                  assert(tag != EXTref);
                  global.definitions.RootClass
                } else if (tag == EXTMODCLASSref) {
                  owner.info.decl(name).moduleClass
                } else {
                  owner.info.decl(name)
                }
              entries(i) = sym;
              if (sym == NoSymbol)
                errorBadSignature(
                  "reference " + (if (name.isTypeName) "type " else "value ") +
                  name.decode + " of " + owner + " refers to nonexisting symbol.");
          case _ =>
            assert(isSymbolEntry(i));
            val name = readNameRef();
            if (settings.debug.value) global.log("reading " + name + " at " + i);
            val owner = readSymbolRef();
            if (entries(i) != null)
              sym = entries(i).asInstanceOf[Symbol]
            else {
              val pflags = readNat();
              val inforef = readNat();
              var syminfo: Type = null;
              tag match {
                case TYPEsym =>
                  sym = owner.newAbstractType(Position.NOPOS, name);
                  entries(i) = sym;
		  syminfo = new LazyBoundsTypeRef(
		    inforef, readNat(), (pflags & Modifiers.VIEWBOUND) != 0)
                case ALIASsym =>
                  sym = owner.newAliasType(Position.NOPOS, name);
                  entries(i) = sym;
                  syminfo = new LazyTypeRef(inforef, sym);
                  val constr = readSymbolRef();
                  if (!constr.typeParams.isEmpty)
                    syminfo = new LazyPolyType(
                      constr.typeParams map (.cloneSymbol(sym)),
                      syminfo);
                case CLASSsym =>
                  sym =
                    if ((pflags & Modifiers.MODUL) != 0) {
                      val modulesym = readSymbolRef();
                      modulesym.moduleClass
                    } else if (name == classRoot.name && owner == classRoot.owner) {
                      if (settings.debug.value) global.log("overwriting " + classRoot);
                      classRoot
                    } else {
                      owner.newClass(Position.NOPOS, name)
                    }
                  entries(i) = sym;
                  syminfo = new LazyTypeRef(inforef, sym);
                  sym.setTypeOfThis(new LazyTypeRef(readNat(), sym));
                  val constr = readSymbolRef();
                  if (!constr.typeParams.isEmpty)
                    syminfo = new LazyPolyType(
                      constr.typeParams map (.cloneSymbol(sym)),
                      syminfo);
                case VALsym =>
                  sym =
                    if (name == moduleRoot.name && owner == moduleRoot.owner) {
                      if (settings.debug.value) global.log("overwriting " + moduleRoot);
                      moduleRoot
                    } else if ((pflags & Modifiers.MODUL) != 0) {
                      owner.newModule(Position.NOPOS, name)
                    } else {
                      if (name == nme.CONSTRUCTOR && bp < end) readSymbolRef();
                      owner.newValue(Position.NOPOS, name);
                    }
                    if (sym.isModule) {
                      val clazz = readSymbolRef();
                      assert(clazz == sym.moduleClass, sym)
                    }
                    syminfo = new LazyTypeRef(inforef, sym);
                case _ =>
                  errorBadSignature("");
              }
              sym.setFlag(transFlags(pflags));
              sym.setInfo(syminfo);
              enterSymbol(sym);
            }
        }
      }
    }
    sym
  }

  def readSymbolRef(): Symbol = getSymbol(readNat());

  def readSymbolRefs(end: int): List[Symbol] =
    if (bp == end) List()
    else {
      val symref = readNat();
      if (!isSymbolEntry(symref)) List()
      else getSymbol(symref) :: readSymbolRefs(end)
    }

  def enterSymbol(sym: Symbol): unit =
    if (sym.owner.isClass && !sym.isModuleClass) {
      if (settings.debug.value) global.log("entering " + sym + ":" + sym.tpe + " in " + sym.owner);//debug
      val scope = sym.owner.info.decls;
      val other = scope.lookup(sym.name);
      if (other != sym) {
        sym.info match {
          case OverloadedType(alts) => alts foreach scope.enter
          case _ => scope.enter(sym)
        }
      }
    }

  def getType(i: int, owner: Symbol): Type = {
    var tpe = entries(i).asInstanceOf[Type];
    if (tpe == null) {
      at(index(i)) {
	val tag = readByte();
	val end = readNat() + bp;
	tpe = tag match {
	  case NOtpe =>
	    NoType
	  case NOpre =>
	    NoPrefix
	  case THIStpe =>
	    ThisType(readSymbolRef())
	  case SINGLEtpe =>
	    singleType(readTypeRef(owner), readSymbolRef());
	  case CONSTANTtpe =>
            ConstantType(readTypeRef(owner), readConstantRef())
	  case TYPEREFtpe =>
            // create a type-ref as found, without checks or rebinds
	    new ExtTypeRef(readTypeRef(owner), readSymbolRef(), readTypeRefs(end, owner))
	  case COMPOUNDtpe =>
            val isCompoundSym = readByte() != 0;
            val ctOwner = if (isCompoundSym) readSymbolRef() else null;
            val ctClassRef = readNat();
            val ctClass: Symbol =
	      if (isCompoundSym) entries(ctClassRef).asInstanceOf[Symbol]
	      else getSymbol(ctClassRef);
	    val parents = readTypeRefs(end, owner);
	    if (ctClass == null)
	      refinedType(intersectionType(parents), ctOwner)
	    else if (isCompoundSym)
	      new ExtRefinedType(intersectionType(parents), new Scope, ctClass)
	    else
	      ClassInfoType(parents, new Scope, ctClass);
	  case METHODtpe =>
	    val restype = readTypeRef(owner);
            val argtps = at(bp)(readTypeRefs(end, owner)) map { argtp =>
              val flags = getFlags(readNat());
              if ((flags & DEFflag) != 0)
                PolyType(List(), argtp)
              else if ((flags & REPEATEDflag) != 0)
                appliedType(definitions.RepeatedParamClass.tpe, List(argtp))
              else
                argtp
            }
            MethodType(argtps, restype)
	  case POLYtpe =>
            val restype = readTypeRef(owner);
	    PolyType(readSymbolRefs(end), restype)
	  case OVERLOADEDtpe =>
            val alts = readSymbolRefs(end);
            readTypeRefs(end, NoSymbol);
            OverloadedType(alts)
	  case FLAGGEDtpe =>
	    readNat(); // skip flags
	    readTypeRef(owner)
	  case _ =>
	    errorBadSignature("");
	}
	if (tag != METHODtpe) entries(i) = tpe;
      }
    }
    tpe
  }

  def readTypeRef(owner: Symbol): Type = getType(readNat(), owner);

  def readTypeRefs(end: int, owner: Symbol): List[Type] =
    if (bp == end) List()
    else {
      val typeref = readNat();
      if (!isTypeEntry(typeref)) List()
      else getType(typeref, owner) :: readTypeRefs(end, owner)
    }

  def getFlags(i: int): int =
    at(index(i)) {
      val tag = readByte();
      val end = readNat() + bp;
      if (tag == FLAGGEDtpe) readNat() else 0
    }

  def getConstant(i: int): Any = {
    var value = entries(i);
    if (value == null) {
      at(index(i)) {
        val tag = readByte();
        val len = readNat();
        value = tag match {
          case LITERALunit    => ()
          case LITERALboolean => if (readByte() == 0) false else true
          case LITERALbyte    => readLong(len).asInstanceOf[byte]
          case LITERALshort   => readLong(len).asInstanceOf[short]
          case LITERALchar    => readLong(len).asInstanceOf[char]
          case LITERALint     => readLong(len).asInstanceOf[int]
          case LITERALlong    => readLong(len)
          case LITERALfloat   => Float.intBitsToFloat(readLong(len).asInstanceOf[int])
          case LITERALdouble  => Double.longBitsToDouble(readLong(len))
          case LITERALstring  => readNameRef().toString()
          case LITERALnull    => null
          case _              => errorBadSignature("bad constant tag: " + tag)
        }
        entries(i)= value
      }
    }
    value
  }

  def readConstantRef(): Any = getConstant(readNat());

  def errorBadSignature(msg: String) =
    throw new RuntimeException("malformed Scala signature at " + bp + "; " + msg);

  private case class OverloadedType(alts: List[Symbol]) extends Type;

  private class LazyTypeRef(tpref: int, owner: Symbol) extends LazyType {
    override def complete(sym: Symbol): unit =
      sym.setInfo(getType(tpref, owner));
  }

  private class LazyBoundsTypeRef(hiref: int, loref: int, vubound: boolean) extends LazyType {
    override def complete(sym: Symbol): unit = {
      val hi = if (vubound) definitions.AnyClass.tpe else getType(hiref, sym);
      val vu = if (vubound) getType(hiref, sym) else definitions.AnyClass.tpe;
      val lo = getType(loref, sym);
      sym.setInfo(TypeBounds(lo, hi, vu))
    }
  }

// --- print symbl files -------------------------------------------------

  private def tag2string(tag: int): String = tag match {
    case TERMname => "TERMname";
    case TYPEname => "TYPEname";
    case NONEsym => "NONEsym";
    case TYPEsym => "TYPEsym";
    case ALIASsym => "ALIASsym";
    case CLASSsym => "CLASSsym";
    case VALsym => "VALsym";
    case EXTref => "EXTref";
    case EXTMODCLASSref => "EXTMODCLASSref";
    case NOtpe => "NOtpe";
    case THIStpe => "THIStpe";
    case SINGLEtpe => "SINGLEtpe";
    case TYPEREFtpe => "TYPEREFtpe";
    case CONSTANTtpe => "CONSTANTtpe";
    case COMPOUNDtpe => "COMPOUNDtpe";
    case METHODtpe => "METHODtpe";
    case POLYtpe => "POLYtpe";
    case OVERLOADEDtpe => "OVERLOADEDtpe";
    case UNBOXEDtpe => "UNBOXEDtpe";
    case UNBOXEDARRAYtpe => "UNBOXEDARRAYtpe";
    case FLAGGEDtpe => "FLAGGEDtpe";
    case ERRORtpe => "ERRORtpe";
    case LITERALunit => "LITERALunit";
    case LITERALboolean => "LITERALboolean";
    case LITERALbyte => "LITERALbyte";
    case LITERALshort => "LITERALshort";
    case LITERALchar => "LITERALchar";
    case LITERALint => "LITERALint";
    case LITERALlong => "LITERALlong";
    case LITERALfloat => "LITERALfloat";
    case LITERALdouble => "LITERALdouble";
    case LITERALstring => "LITERALstring";
    case LITERALnull => "LITERALnull";
    case LITERALzero => "LITERALzero";
    case _ => "***BAD TAG***(" + tag + ")";
  }

  def printAll(out: PrintStream): unit = {
    out.println("symbl attribute for " + classRoot + ":");
    for (val i <- Iterator.range(0, index.length)) {
      out.print(i + "," + index(i) + ": ");
      bp = index(i);
      val tag = readByte();
      out.print(tag2string(tag));
      val len = readNat();
      val end = len + bp;
      out.print(" " + len);
      tag match {
	case TERMname | TYPEname =>
	  out.print(" " + UTF8Codec.decode(bytes, bp, len));
	  bp = end;
	case NONEsym =>
        case TYPEsym | ALIASsym | CLASSsym | VALsym =>
	  out.print(" " + readNat()); //name
	  out.print(" " + readNat()); //owner
	  out.print(" " + Integer.toHexString(readNat())); //flags
	  out.print(" " + readNat()); //type
	case FLAGGEDtpe =>
	  out.print(" " + Integer.toHexString(readNat())); //flags
      }
      while (bp < end) out.print(" " + readNat());
      out.println();
    }
  }

  def transFlags(pflags: int): long = {
    var res = 0L;
    if ((pflags & Modifiers.DEFERRED     ) != 0) res = res | Flags.DEFERRED   ;
    if ((pflags & Modifiers.FINAL        ) != 0) res = res | Flags.FINAL      ;
    if ((pflags & Modifiers.PRIVATE      ) != 0) res = res | Flags.PRIVATE    ;
    if ((pflags & Modifiers.PROTECTED    ) != 0) res = res | Flags.PROTECTED  ;
    if ((pflags & Modifiers.SEALED       ) != 0) res = res | Flags.SEALED     ;
    if ((pflags & Modifiers.OVERRIDE     ) != 0) res = res | Flags.OVERRIDE   ;
    if ((pflags & Modifiers.CASE         ) != 0) res = res | Flags.CASE       ;
    if ((pflags & Modifiers.ABSTRACT     ) != 0) res = res | Flags.ABSTRACT   ;
    if ((pflags & Modifiers.SYNTHETIC    ) != 0) res = res | Flags.SYNTHETIC  ;
    if ((pflags & Modifiers.DEPRECATED   ) != 0) res = res | Flags.DEPRECATED ;
    if ((pflags & Modifiers.JAVA         ) != 0) res = res | Flags.JAVA       ;
    if ((pflags & Modifiers.MODUL        ) != 0) res = res | Flags.MODULE     ;
    if ((pflags & Modifiers.MUTABLE      ) != 0) res = res | Flags.MUTABLE    ;
    if ((pflags & Modifiers.PARAM        ) != 0) res = res | Flags.PARAM      ;
    if ((pflags & Modifiers.INITIALIZED  ) != 0) res = res | Flags.INITIALIZED;
    if ((pflags & Modifiers.LOCKED       ) != 0) res = res | Flags.LOCKED     ;
    if ((pflags & Modifiers.ACCESSED     ) != 0) res = res | Flags.ACCESSED   ;
    if ((pflags & Modifiers.SELECTOR     ) != 0) res = res | Flags.SELECTOR   ;
    if ((pflags & Modifiers.PACKAGE      ) != 0) res = res | Flags.PACKAGE    ;
    if ((pflags & Modifiers.STABLE       ) != 0) res = res | Flags.STABLE     ;
    if ((pflags & Modifiers.CAPTURED     ) != 0) res = res | Flags.CAPTURED   ;
    if ((pflags & Modifiers.INCONSTRUCTOR) != 0) res = res | Flags.INCONSTRUCTOR;
    if ((pflags & Modifiers.PARAMACCESSOR) != 0) res = res | Flags.PARAMACCESSOR;
    if ((pflags & Modifiers.ACCESSOR     ) != 0) res = res | Flags.ACCESSOR   ;
    if ((pflags & Modifiers.BRIDGE       ) != 0) res = res | Flags.BRIDGE     ;
    if ((pflags & Modifiers.LIFTED       ) != 0) res = res | Flags.LIFTED     ;
    if ((pflags & Modifiers.ALTERNATIVE  ) != 0) res = res | Flags.ALTERNATIVE;
    if ((pflags & Modifiers.INTERFACE    ) != 0) res = res | Flags.TRAIT      ;
    if ((pflags & Modifiers.TRAIT        ) != 0) res = res | Flags.TRAIT      ;
    if ((pflags & Modifiers.COVARIANT    ) != 0) res = res | Flags.COVARIANT  ;
    if ((pflags & Modifiers.CONTRAVARIANT) != 0) res = res | Flags.CONTRAVARIANT;
    res
  }
}
}
*/

File: sources/scalac/symtab/classfile/UnPickle.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
\*                                                                      */

// $Id$

package scalac.symtab.classfile;

import java.util.HashMap;
import scalac.*;
import scalac.util.*;
import ch.epfl.lamp.util.Position;
import scalac.symtab.*;
import Symbol.*;
import Type.*;

public class UnPickle implements Kinds, Modifiers, EntryTags {

/***************************************************
 * Symbol table attribute format:
 *   Symtab         = nentries_Nat {Entry}
 *   Entry          = TERMNAME len_Nat NameInfo
 *                  | CONSTRNAME len_Nat NameInfo
 *                  | TYPENAME len_Nat NameInfo
 *                  | NONEsym len_Nat
 *                  | TYPEsym len_Nat SymbolInfo lobound_Ref
 *                  | ALIASsym len_Nat SymbolInfo
 *                  | CLASSsym len_Nat SymbolInfo thistype_Ref constrsym_Ref
 *                  | MODULEsym len_Nat SymbolInfo classsym_Ref
 *                  | VALsym len_Nat SymbolInfo [classsym_Ref]
 *                  | EXTsym len_Nat name_Ref [owner_Ref]
 *                  | NOtpe len_Nat
 *                  | THIStpe len_Nat sym_Ref
 *                  | SINGLEtpe len_Nat type_Ref sym_Ref
 *                  | TYPEREFtpe len_Nat type_Ref sym_Ref {targ_Ref}
 *                  | COMPOUNDtpe len_Nat sym_Ref {tpe_Ref} {sym_Ref}
 *                  | METHODtpe len_Nat tpe_Ref {tpe_Ref}
 *                  | POLYTtpe len_Nat tpe_Ref {sym_Ref}
 *                  | OVERLOADEDtpe len_Nat {sym_Ref} {tpe_Ref}
 *                  | FLAGGEDtype len_Nat flags_Nat tpe_Ref
 *   SymbolInfo     = name_Ref owner_Ref flags_Nat info_Ref
 *   NameInfo       = <character sequence of length len_Nat in Utf8 format>
 *   Ref            = Nat
 *
 *   len is remaining length after `len'.
 */
    static final boolean debug = false;

    Symbol classroot;
    Symbol moduleroot;
    byte[] bytes;
    int bp;
    Name sourceName;
    int[] index;
    Object[] entries;
    int paramFlags;

    UnPickle(Symbol root, byte[] data, Name sourceName) {
	assert root.rawInfoAt(Symbol.FIRST_ID) instanceof LazyType;
	//System.out.println("unpickle " + root);//DEBUG
	this.classroot = root;
	this.moduleroot = classroot.module();
	this.bytes = data;
	this.bp = 0;
	this.sourceName = sourceName;
	index = new int[readNat()];
	for (int i = 0; i < index.length; i++) {
	    index[i] = bp;
	    bp++;
	    bp = readNat() + bp;
	}
	entries = new Object[index.length];
	if (debug) {
	    for (int i = 0; i < index.length; i++) {
		System.out.print(i + "," + index[i] + ": ");
		bp = index[i];
		System.out.print(readByte() + " ");
		int len = readNat();
		System.out.print(len + " ");
		for (int j = 0; j < len; j++)
		    System.out.print(readByte() + " ");
		System.out.println();
	    }
	}
	for (int i = 0; i < index.length; i++) {
	    if (isSymbolEntry(i)) getSymbol(i);
	}
	if (root.rawInfoAt(Symbol.FIRST_ID) instanceof LazyType)
	    throw new BadSignature(this, "it does not define " + root);
    }

    Type setOwner(Type tp, Symbol owner) {
	switch (tp) {
	case PolyType(Symbol[] tparams, Type restpe):
	    Type restpe1 = setOwner(restpe, owner);
	    if (restpe1 == restpe) return tp;
	    else return Type.PolyType(tparams, restpe1);
	case MethodType(Symbol[] params, Type restpe):
	    Symbol[] params1 = params;
	    for (int i = 0; i < params.length; i++) {
		if (params[i].owner() != owner) {
		    if (params1 == params && params[i].owner() != Symbol.NONE) {
			params1 = new Symbol[params.length];
			System.arraycopy(params, 0, params1, 0, i);
		    }
		    params1[i].setOwner(owner);
		}
	    }
	    Type restpe1 = setOwner(restpe, owner);
	    if (params1 == params && restpe1 == restpe) return tp;
	    else return Type.MethodType(params1, restpe1);
	default:
	    return tp;
	}
    }

    int readByte() {
	return bytes[bp++];
    }

    int readNat() {
	int b;
	int x = 0;
	do {
	    b = readByte();
	    x = (x << 7) + (b & 0x7f);
	} while ((b & 0x80) != 0);
	return x;
    }

    boolean isTypeEntry(int i) {
	int tag = bytes[index[i]];
	return (firstTypeTag <= tag && tag <= lastTypeTag);
    }

    boolean isSymbolEntry(int i) {
	int tag = bytes[index[i]];
	return (firstSymTag <= tag && tag <= lastSymTag);
    }

    Name getName(int n) {
	if (entries[n] == null) {
	    int savedBp = bp;
	    bp = index[n];
	    int tag = bytes[bp++];
	    int len = readNat();
	    Name name = Name.fromAscii(bytes, bp, len);
	    switch (tag) {
	    case TERMname  : entries[n] = name; break;
	    case CONSTRname: entries[n] = name.toConstrName(); break;
	    case TYPEname  : entries[n] = name.toTypeName(); break;
	    default: throw new BadSignature(this);
	    }
	    bp = savedBp;
	}
	return (Name) entries[n];
    }

    Name readNameRef() {
	return getName(readNat());
    }

    Symbol getSymbol(int n) {
	if (entries[n] == null) {
	    int savedBp = bp;
	    bp = index[n];
	    int tag = bytes[bp++];
	    int end = readNat() + bp;
	    Symbol sym;
	    Symbol owner;
	    switch (tag) {
	    case NONEsym:
		entries[n] = sym = Symbol.NONE;
		break;
	    case EXTsym:
		Name name = readNameRef();
		if (bp == end) {
		    owner = Global.instance.definitions.ROOT_CLASS;
		} else {
		    assert bp < end;
		    owner = readSymbolRef();
		}
		if (name.length() == 0 && owner == Symbol.NONE) {
		    entries[n] = sym = Global.instance.definitions.ROOT_CLASS;
		} else {
		    entries[n] = sym = owner.info().lookupNonPrivate(name);
		}
		if (sym.kind == NONE) {
		    throw new BadSignature(this,
			"reference " + name + " of " + owner +
			" refers to nonexisting symbol.");
		}
		break;
	    default:
		assert isSymbolEntry(n) : n;
		Name name = readNameRef();
		owner = readSymbolRef();
		if (entries[n] == null) {
		    int flags = readNat();
		    int inforef = readNat();
		    switch (tag) {
		    case TYPEsym:
			entries[n] = sym = new AbsTypeSymbol(
			    Position.NOPOS, name, owner, flags);
			sym.setInfo(getType(inforef));
			sym.setLoBound(readTypeRef());
			break;

		    case ALIASsym:
			entries[n] = sym = new TypeSymbol(
			    ALIAS, Position.NOPOS, name, owner, flags);
			sym.setInfo(getType(inforef));
			break;

		    case CLASSsym:
			entries[n] = sym = new ClassSymbol(
			    Position.NOPOS, name, owner, flags);
			if (name == classroot.name && owner == classroot.owner()) {
			    sym.copyTo(classroot);
			    entries[n] = sym = classroot;
			}
			sym.setInfo(getType(inforef));
			sym.setTypeOfThis(readTypeRef());
			Symbol constructor = readSymbolRef(); //will install itself!
			break;

		    case MODULEsym:
			entries[n] = sym = TermSymbol.newModule(
			    Position.NOPOS, name, owner, flags, (ClassSymbol) readSymbolRef());
			if (name == moduleroot.name && owner == moduleroot.owner()) {
			    sym.copyTo(moduleroot);
			    entries[n] = sym = moduleroot;
			}
			sym.setInfo(getType(inforef));
			break;

		    case VALsym:
			if (bp < end) {
			    assert name.isConstrName();
			    Symbol clazz = readSymbolRef();
			    entries[n] = sym = clazz.constructor();
			    TermSymbol.newConstructor(clazz, flags).copyTo(sym);
			} else {
			    entries[n] = sym = new TermSymbol(
				Position.NOPOS, name, owner, flags);
			}
			sym.setInfo(setOwner(getType(inforef), sym));
			break;

		    default:
			throw new BadSignature(this);
		    }

		    if (owner.kind == CLASS &&
			!(sym == classroot ||
			  sym.kind == CLASS && (sym.flags & MODUL) != 0 ||
			  (sym.isPrimaryConstructor() &&
			   (sym.primaryConstructorClass().flags & MODUL) != 0)))
			owner.members().enter(sym);
		}
	    }
	    bp = savedBp;
	}
	return (Symbol) entries[n];
    }

    Symbol readSymbolRef() {
	return getSymbol(readNat());
    }

    Symbol[] readSymbolRefs(int end) {
	return readSymbolRefs(0, end);
    }

    Symbol[] readSymbolRefs(int nread, int end) {
	if (bp == end) {
	    return new Symbol[nread];
	} else {
	    assert bp < end;
	    int ref = readNat();
	    if (isSymbolEntry(ref)) {
		Symbol s = getSymbol(ref);
		Symbol[] ss = readSymbolRefs(nread+1, end);
		ss[nread] = s;
		return ss;
	    } else {
		return new Symbol[nread];
	    }
	}
    }

    Type getType(int n) {
	if (entries[n] == null) {
	    int savedBp = bp;
	    bp = index[n];
	    int tag = bytes[bp++];
	    int end = readNat() + bp;
	    Type tpe;
	    switch (tag) {
	    case NOtpe:
		tpe = Type.NoType;
		break;
	    case THIStpe:
		tpe = Type.ThisType(readSymbolRef());
		break;
	    case SINGLEtpe:
		tpe = Type.singleType(readTypeRef(), readSymbolRef());
		break;
	    case TYPEREFtpe:
		tpe = Type.TypeRef(
		    readTypeRef(), readSymbolRef(), readTypeRefs(end));
		break;
	    case COMPOUNDtpe:
		Symbol clazz = readSymbolRef();
		Type[] parents = readTypeRefs(end);
		Scope members = new Scope(readSymbolRefs(end));
		tpe = Type.compoundType(parents, members, clazz);
		break;
	    case METHODtpe:
		Type restype = readTypeRef();
		int bp1 = bp;
		Type[] argtypes = readTypeRefs(end);
		int[] flags = new int[argtypes.length];
		bp = bp1;
		readFlags(flags);
		Symbol[] params = new Symbol[argtypes.length];
		for (int i = 0; i < argtypes.length; i++) {
		    params[i] = new TermSymbol(
			Position.NOPOS, Name.fromString("$" + i),
			Symbol.NONE, PARAM | flags[i]);
		    params[i].setInfo(argtypes[i]);
		}
		tpe = Type.MethodType(params, restype);
		break;
	    case POLYtpe:
		Type restype = readTypeRef();
		tpe = Type.PolyType(readSymbolRefs(end), restype);
		break;
	    case OVERLOADEDtpe:
		Symbol[] alts = readSymbolRefs(end);
		Type[] alttypes = readTypeRefs(end);
		for (int i = 0; i < alts.length; i++)
		    alttypes[i] = setOwner(alttypes[i], alts[i]);
		tpe = Type.OverloadedType(alts, alttypes);
		break;
	    case FLAGGEDtpe:
		paramFlags = getFlags(readNat());
		return readTypeRef();
	    default:
		throw new BadSignature(this);
	    }
	    entries[n] = tpe;
	    bp = savedBp;
	}
	return (Type) entries[n];
    }

    Type readTypeRef() {
	return getType(readNat());
    }

    Type[] readTypeRefs(int end) {
	return readTypeRefs(0, end);
    }

    Type[] readTypeRefs(int nread, int end) {
	if (bp == end) {
	    return new Type[nread];
	} else {
	    assert bp < end;
	    int ref = readNat();
	    if (isTypeEntry(ref)) {
		Type t = getType(ref);
		Type[] ts = readTypeRefs(nread + 1, end);
		ts[nread] = t;
		return ts;
	    } else {
		return new Type[nread];
	    }
	}
    }

    void readFlags(int[] flags) {
	for (int i = 0; i < flags.length; i++)
	    flags[i] = getFlags(readNat());
    }

    int getFlags(int n) {
	int savedBp = bp;
	bp = index[n];
	int tag = bytes[bp++];
	int end = readNat() + bp;
	int flags = (tag == FLAGGEDtpe) ? decodeFlags(readNat()) : 0;
	bp = savedBp;
	return flags;
    }

    private static int decodeFlags(int n) {
	int flags = 0;
	if ((n & COVARflag) != 0) flags |= COVARIANT;
	if ((n & CONTRAVARflag) != 0) flags |= CONTRAVARIANT;
	if ((n & REPEATEDflag) != 0) flags |= REPEATED;
	if ((n & DEFflag) != 0) flags |= DEF;
	return flags;
    }

    public static class BadSignature extends java.lang.Error {
	public BadSignature(UnPickle outer, String msg) {
	    super("symbol data " + outer.sourceName +
		  " could not be loaded; " + msg);
	}
	public BadSignature(UnPickle outer) {
	    this(outer, "malformed signature at " + outer.bp);
	}
    }
}

